static/js/publisher/form/banner.test.js

```diff
@@ -0,0 +1,167 @@
+import React from "react";
+import { render, fireEvent } from "react-testing-library";
+
+import Banner from "./banner";
+
+const backgroundSelector = ".p-market-banner__image img";
+
+function renderBanner(bannerImage, updateState, restrictions) {
+  if (bannerImage && updateState) {
+    return render(
+      <Banner bannerImage={bannerImage} updateImageState={updateState} />
+    );
+  }
+  if (bannerImage && updateState) {
+    return render(
+      <Banner bannerImage={bannerImage} updateImageState={updateState} />
+    );
+  }
+  if (updateState) {
+    return render(<Banner updateImageState={updateState} />);
+  }
+  if (bannerImage) {
+    return render(<Banner bannerImage={bannerImage} />);
+  }
+
+  if (restrictions) {
+    return render(
+      <Banner
+        restrictions={{
+          accept: ["image/jpeg"]
+        }}
+      />
+    );
+  }
+
+  return render(<Banner />);
+}
+
+describe("Banner", () => {
+  it("should render with no props", () => {
+    const { container } = renderBanner();
+
+    expect(
+      container.querySelectorAll(".p-market-banner__image-holder").length
+    ).toEqual(1);
+  });
+
+  describe("background image", () => {
+    it("should show the background image if set at render", () => {
+      const { container } = renderBanner({ url: "banner.png" });
+
+      expect(container.querySelector(backgroundSelector).src).toEqual(
+        "banner.png"
+      );
+    });
+
+    it("should set the background if image added", done => {
+      const { container } = renderBanner();
+
+      window.URL = {
+        createObjectURL: () => {
+          return "banner.png";
+        }
+      };
+
+      const fileContents = "test";
+      const file = new File([fileContents], "banner.png", {
+        type: "image/png"
+      });
+
+      const input = container.querySelector(`[name="banner-image"]`);
+
+      Object.defineProperty(input, "files", {
+        value: [file]
+      });
+
+      fireEvent.change(input);
+
+      setTimeout(() => {
+        expect(
+          container.querySelectorAll(".p-notification--negative").length
+        ).toEqual(0);
+        expect(container.querySelector(backgroundSelector).src).toEqual(
+          "banner.png"
+        );
+        done();
+      }, 500);
+    });
+
+    it("should show an error if a restriction isn't met", done => {
+      const { container } = renderBanner(null, null, true);
```

Comment:
Extracting `renderBanner` resulted in tests being harder to read.
I have no idea what `null, null, true` means. Just rendering `<Banner restrictions={{ accept: ["image/jpeg"] }} />` here inline would make it much more clear. I don't see such specific banner configurations to be reused much to justify having extracted renderBanner function (and component props obscured in function params).

Reply:
Yeah, the tests were more complex before, I've reduced them down a lot because much of the functionality is handled by the fileInput component now!